### PR TITLE
Implement arithmetic overflow changes

### PR DESCRIPTION
**Pull request by pnkfelix**
Rebase and follow-through on work done by @cmr and @aatch.

Implements most of rust-lang/rfcs#560. Errors encountered from the checks during building were fixed.

The checks for division, remainder and bit-shifting have not been implemented yet.

See also PR #20795, as well as the tracking issue #22020

cc @Aatch ; cc @nikomatsakis

### CODE REVIEW

**Comment by pnkfelix** [comment]
(I'm actually still running `make check-stage1 && make check-stage2` right now; I'll be reporting back when they finish.  Also, not necessarily all of the review notes from #20795 were addressed; I will go back over that PR and see which notes are still relevant.)

**Comment by pnkfelix** [comment]
current PR gets to `run: x86_64-apple-darwin/stage1/test/stdtest-x86_64-apple-darwin` and then hits overflow panics in `collections::hash::map::test_map`, `num::strconv::tests`, `num::tests`, `rand::test`, `u16::tests`,  `u32::tests`, `u64::tests`, `u8::tests`, `uint::tests`, `usize::tests`.  On skimming the names of tests that fail, most overflows seem likely intentional and easily addressed.

**Comment by pnkfelix** [comment]
okay, now `stdtest` finshes, and the next problem we hit is in `libserialize` tests: `json::tests::test_read_number`

**Comment by pnkfelix** [comment]
Now `stdtest` and `serializetest` both finish, and we hit the big one: `collectionstest`! : 

```
failures:
    btree::map::test::test_basic_large
    btree::map::test::test_iter_mixed
    btree::map::test::test_range
    btree::map::test::test_range_1000
    btree::map::test::test_range_small
    ring_buf::tests::test_clone
    ring_buf::tests::test_insert
    ring_buf::tests::test_mut_rev_iter
    ring_buf::tests::test_mut_rev_iter_wrap
    ring_buf::tests::test_param_int
    ring_buf::tests::test_param_reccy
    ring_buf::tests::test_param_taggy
    ring_buf::tests::test_param_taggypar
    ring_buf::tests::test_remove
    ring_buf::tests::test_rev_iter
    ring_buf::tests::test_simple
    ring_buf::tests::test_split_off
    ring_buf::tests::test_swap_front_back_remove
    slice::tests::test_element_swaps
    slice::tests::test_permutations
```

(looking)

**Comment by pnkfelix** [comment]
Addressed `ring_buf`, `slice`, and follow-on problems in `librand`.

The `btree` problem above seems like it might have been a legitimate bug (it seems like the `size_hint` in an iterator gets corrupted, and the tests are not using corrupted data), but I cannot seem to expose this problem in the playpen, so either its been fixed in the meantime, or my analysis/hypothesis is wrong.  Anyway, I have not checked in my local workaround for `btree`, because it bears further investigation.

Next up:
```
run: x86_64-apple-darwin/stage1/test/coretesttest-x86_64-apple-darwin
...
failures:
    iter::test_random_access_rev
    num::test::test_int_from_str_overflow
```

**Comment by pnkfelix** [comment]
I think I've addressed the two problems above, but now I need to rebuild core.

Q: Was this a bug in `iter` ?  (Or am I injecting one?)  I cannot tell from the contract of `RandomAccessIter` whether the underflow wrap-around means that it should always be returning `None` in this case, or if some iterators (e.g. ones that repeat infinitely) would/should attempt to return a valid `Some(thing)` for a `usize::MAX` index.

```
--- INDEX:/src/libcore/iter.rs
+++ WORKDIR:/src/libcore/iter.rs
@@ -1127,7 +1127,11 @@
 #[unstable(feature = "core", reason = "trait is experimental")]
 impl<I> RandomAccessIterator for Rev<I> where I: DoubleEndedIterator + RandomAccessIterator {
     #[inline]
     fn indexable(&self) -> usize { self.iter.indexable() }
     #[inline]
     fn idx(&mut self, index: usize) -> Option<<I as Iterator>::Item> {
         let amt = self.indexable();
-        self.iter.idx(amt - index - 1)
+        if amt > index {
+            self.iter.idx(amt - index - 1)
+        } else {
+            None
+        }
     }
 }

```

**Comment by alexcrichton** [comment]
@pnkfelix that looks like a legitimate bug fix to me

**Comment by pnkfelix** [comment]
Okay, `libcoretest` works now (though I still need to look into the aforementioned `btree` issue).

Now I get to:
```
run doc-crate-std [x86_64-apple-darwin]
...
failures:
    rand::random_0
```

**Comment by pnkfelix** [comment]
Well, this is a funny one; it obviously might overflow:
```
/// ```
/// use std::rand;
///
/// let x = rand::random();
/// println!("{}", 2u8 * x);
///
/// let y = rand::random::<f64>();
/// println!("{}", y);
///
/// if rand::random() { // generates a boolean
///     println!("Better lucky than good!");
/// }
/// ```
```

**Comment by pnkfelix** [comment]
Current state now:
```
run rpass [x86_64-apple-darwin]: x86_64-apple-darwin/stage2/bin/compiletest
...
failures:
    [run-pass] run-pass/enum-discr.rs
    [run-pass] run-pass/issue-14021.rs
```

**Comment by pnkfelix** [comment]
Okay, we get past `run-pass` now.  (Which was ... easier than I expected.  I wonder if there are no run-pass tests that check handling of constant expressions that overflow...)

Next, compile-fail, we do surprisingly well there too (probably too well, again I say):

```
run cfail [x86_64-apple-darwin]: x86_64-apple-darwin/stage2/bin/compiletest
[...]
failures:
    [compile-fail] compile-fail/lifetime-elision-return-type-requires-explicit-lifetime.rs

test result: FAILED. 1583 passed; 1 failed; 20 ignored; 0 measured
```

**Comment by pnkfelix** [comment]
Now we get past `compile-fail`.  Next up, `run-make`:

```
run cfail-full [x86_64-apple-darwin]: x86_64-apple-darwin/stage2/bin/compiletest
[...]
maketest: alloc-extern-crates
maketest: allow-non-lint-warnings-cmdline
maketest: allow-warnings-cmdline-stability
maketest: c-dynamic-dylib
maketest: c-dynamic-rlib
maketest: c-link-to-rust-dylib
maketest: c-link-to-rust-staticlib
maketest: c-static-dylib
maketest: c-static-rlib
maketest: cannot-read-embedded-idents
maketest: codegen-options-parsing
maketest: compiler-lookup-paths-2
maketest: compiler-lookup-paths
maketest: crate-data-smoke
maketest: crate-name-priority
maketest: dep-info-spaces
[...]
maketest: save-analysis
maketest: sepcomp-cci-copies
```

and then `rustc` presumably hits an internal arithmetic overflow while trying to build that last test.

**Comment by pnkfelix** [comment]
(I can't figure out why the `sepcomp` tests are failing for me.  For now I'm going to skip them, under the assumption that something else i did to the `rustc`, e.g. stuff to turn off the inlining attribute for debugging, is the cause here and that this will fix itself when I remove those changes and test against a clean build.)

**Comment by pnkfelix** [comment]
* I figured out why I had not been able to expose an underflow bug in `btree`: AFAICT, we only attempt to decrement `0usize` during the range iterators, which explicitly say that they ignore the `size` field (by not passing through the `fn size_method` method to their inner `AbsIter` the way that the other adapters do).  I addressed this by setting the unused `size` field to `usize::MAX` instead of `0`.

 * I changed constant-expression evaluation to use `OverflowingOps` methods rather than blind arithmetic, to be able to handle the subsequent overflow cases rather than panicking in the stage2 compiler.

 * I revamped the error-handling during constant-expression evaluation.

 * I found some more sources of over- and underflow, and fixed them.  No obvious bugs exposed by this, though I am at best suspicious of our constant evaluatior.

 * I revised a number of places that either ICE'd or issued fatal errors to instead issue a normal error, to allow computation to proceed.  This was a little bit too aggressive; the way I did it here causes blowup in the amount of redundant error reporting.

However, all together, these changes appear to let me pass through `make check-stage2`!

File: src/librustc_trans/trans/base.rs [hunk 1]
```diff
@@ -3083,6 +3083,12 @@ pub fn trans_crate<'tcx>(analysis: ty::CrateAnalysis<'tcx>)
     let ty::CrateAnalysis { ty_cx: tcx, export_map, reachable, name, .. } = analysis;
     let krate = tcx.map.krate();
 
+    let check_overflow = if let Some(v) = tcx.sess.opts.debugging_opts.force_overflow_checks {
+        v
+    } else {
+        !attr::contains_name(&krate.config[], "ndebug")
```
Comment (pnkfelix) [inline src/librustc_trans/trans/base.rs]: (this may need to be changed to be based on optimization level rather than `ndebug`)

**Comment by pnkfelix** [comment]
@eddyb points out correctly that while this code is catching (and reporting) overflows of 64-bit integer arithmetic in `const_eval.rs`, it does not appear to catch e.g. an attempt to const-eval `255u8 + 1u8`.  I will file such as a separate ticket, since I am guessing that it is a serious P-backcompat-lang wart that we do not want to be stuck with.

File: src/librustc_typeck/astconv.rs [hunk 2]
```diff
@@ -1381,14 +1383,22 @@ pub fn ast_ty_to_ty<'tcx>(this: &AstConv<'tcx>,
...
                         }
                     }
                 }
-                Err(r) => {
-                    span_fatal!(tcx.sess, ast_ty.span, E0250,
-                                "expected constant expr for array length: {}", r);
+                Err(ref r) => {
+                    let subspan  =
+                        ast_ty.span.lo <= r.span.lo && r.span.hi <= ast_ty.span.hi;
+                    span_err!(tcx.sess, ast_ty.span, E0250,
```
Comment (pnkfelix) [inline src/librustc_typeck/astconv.rs]: hmm, I am pretty sure I meant for this to pass in `r.span` here.  (and then you only get the whole `ast_ty.span` if they are actually disconnected).

Comment (pnkfelix) [inline src/librustc_typeck/astconv.rs]: (clearly I need to add a test for this case. :)

File: src/librustc/middle/const_eval.rs [hunk 3]
```diff
@@ -202,15 +204,193 @@ pub fn const_expr_to_pat(tcx: &ty::ctxt, expr: &Expr, span: Span) -> P<ast::Pat>
...
+            NonConstStruct      => "non-constant struct in constant expr".into_cow(),
+            TupleIndexOutOfBounds => "tuple index out of bounds".into_cow(),
+
+            MiscBinaryOp => "bad operands for binary".into_cow(),
+            MiscCatchAll => "unsupported constant expr".into_cow(),
+        }
+    }
+}
+
+fn invalid_op_for_bools(e: &Expr, op: ast::BinOp_) -> ConstEvalErr {
```
Comment (eddyb) [inline src/librustc/middle/const_eval.rs]: This and all the similar functions below could be replaced by a macro which expands to:
```rust
return Err(ConstEvalErr { span: $e.span, kind: $ctor(($args),*) }
```
(with a no-argument form, of course)

File: src/librustc/middle/const_eval.rs [hunk 4]
```diff
@@ -361,10 +538,13 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
...
               ast::BiNe => a != b,
-              _ => return Err("can't do this op on bools".to_string())
+              _ => return Err(invalid_op_for_bools(e, op.node)),
              }))
           }
-          _ => Err("bad operands for binary".to_string())
+          (err @ Err(..), _) |
+          (_, err @ Err(..)) => err,
+
+          _ => Err(misc_binary_op(e)),
```
Comment (eddyb) [inline src/librustc/middle/const_eval.rs]: It's interesting that we do this, but isn't it worse than wrapping the calls to `eval_const_expr_partial` for `a` and `b` in `try!`?

Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: uh, you are referring to the catch-all at the end that returns `Err(misc_binary_op(e))`, right?  I had assumed that this is trying to catch e.g. type-mismatches in the two results ... I do not understand what alternative you are suggesting to that here: the two sides could both hand back `Ok`'s.

Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: (Oh, are you instead referring to the `(err, _) | (_ err)` thing I put in?  *That* could be replaced by `try!'s`, yes, I think so.)

Comment (eddyb) [inline src/librustc/middle/const_eval.rs]: My bad, I was reading the pattern above wrong, I thought it was only matching `(Ok(_), Err(_)) | (Err(_), Ok(_))`.
I guess it's equivalent to using `try!`, so why not do that and get rid of some `Ok` pattern noise?

Comment (pnkfelix) [inline src/librustc/middle/const_eval.rs]: yes I think this is a good idea, I will try it out.

**Comment by eddyb** [comment]
@pnkfelix When I've tried to adjust the results of operations (or just literals? not sure now) for their type, I ended up killing "literal outside of range" warnings (I used the same mechanism `as` uses in `const_eval`, which is much easier than handling it for each possible type in each possible case).
I recall @nikomatsakis agreeing on erroring early in such cases, and I think that with overflow checking finally here, we can start being more aggressive. I would also never infer `256` as an `u8`, but I'm getting ahead of myself.

File: src/librustc/middle/const_eval.rs [hunk 5]
```diff
@@ -441,19 +560,21 @@ pub fn eval_const_expr_partial<'tcx>(tcx: &ty::ctxt<'tcx>,
...
 
-        Err("non-constant struct in constant expr".to_string())
+        signal!(e, NonConstStruct);
       }
-      _ => Err("unsupported constant expr".to_string())
-    }
+      _ => signal!(e, MiscCatchAll)
+    };
+
+    Ok(result)
```
Comment (eddyb) [inline src/librustc/middle/const_eval.rs]: :heart_eyes: Really liking how this function looks now.

**Comment by pnkfelix** [comment]
(I want to give this a whirl on a dedicated linux machine; it is an invasive enough change that I should try to go through `make check-stage2` on the big three (mac/linux/windows) as best I can.)

**Comment by Manishearth** [comment]
I've initiated a build on a linux machine (`check-stage1` for now). Might take a while, this particular machine is behind on llvm and the snapshot.

**Comment by Manishearth** [comment]
Stage1 passed.

**Comment by Manishearth** [comment]
Never mind, I was looking at the wrong build. Still building.

**Comment by Manishearth** [comment]
The backtrace tests are hanging. Confirmed in two runs, I get this [exact same stdout](https://manishearth.pastebin.mozilla.org/8823787) from `check-stage1` barring one reordered test.

`make check-stage1-rpass TESTNAME=backtrace.rs` and `make check-stage1-rpass TESTNAME=backtrace-debuginfo` both hang on a Linux box.

**Comment by Manishearth** [comment]
Both the tests are rather OS-specific (backtrace-debuginfo is mostly Linux-only), which is probably causing this failure.

cc @lifthrasiir

**Comment by lifthrasiir** [comment]
Uh, `backtrace-debuginfo` test *does* rely on the wrapping behavior (which is easy to fix, simply use `i32` counter instead of `u32`), but it should only fail on non-Linux boxes. I don't know why it hangs on Linux.

**Comment by Manishearth** [comment]
https://github.com/rust-lang/rust/blob/master/src/libstd/sys/unix/backtrace.rs#L182 seems to be part of the problem

**Comment by Manishearth** [comment]
Backtraces of the threads spawned by the test: https://manishearth.pastebin.mozilla.org/8823797 https://manishearth.pastebin.mozilla.org/8823798

**Comment by pnkfelix** [comment]
Ah, is there a re-entrancy issue here?  (I.e. attempting to print a backtrace while in the midst of printing a backtrace, and deadlocking in the process?)

**Comment by Manishearth** [comment]
Test fixed, thanks to @lifthrasiir 

Cherry-picking https://github.com/Manishearth/rust/commit/e97a2d7cca4d36e82fa793762353ac909feee8b5 should work.

**Comment by pnkfelix** [comment]
(i am looking into this overflow... it may be 32-bit specific... )

**Comment by pnkfelix** [comment]
>  (i am looking into this overflow... it may be 32-bit specific... )

nuts, it does not reproduce in my VM image.  :(

I could just change the lines in question to be more robust, but I would prefer to observe the effect locally.

**Comment by pnkfelix** [comment]
> I could just change the lines in question to be more robust, but I would prefer to observe the effect locally.

After reviewing the code and thinking it over, I decided the only sane thing to do here would be to act like either over- *or* underflow could be occurring, and to use saturated arithmetic in both cases.  See commit cefd82f

**Comment by pnkfelix** [comment]
> nuts, it does not reproduce in my VM image. :(

Aha!  I have managed to now reproduce it: I had to run `make check` on a *64-bit* VM cross-compiling and running a 32-bit rust binaries!

**Comment by Manishearth** [comment]
Wait, we run our 32bit stuff on 64bit systems? Not 32bit systems/emulators? I can understand the crosscompile on 64bit, but I thought the builders would shove it into qemu or something for testing.

**Comment by Manishearth** [comment]
\o/

**Comment by pnkfelix** [comment]
> >    nuts, it does not reproduce in my VM image. :(
>
> Aha! I have managed to now reproduce it: I had to run make check on a 64-bit VM cross-compiling and running a 32-bit rust binaries!

Mystery solved: It appears it was indeed an overflow triggered from this calcuation:
```
let top_plus_20k = my_stack_top + 20000;
```

**Pull request merged**